File: BC-W3/Java/java/java/Gun/Gun.java
public class Gun implements Cloneable {
    private int amount;
    private int capacity;
    private boolean isReady;
    private String model;
    private int totalShots;
    
    public Gun(String model, int capacity) {
        this.model = model;
        this.capacity = capacity;
        this.amount = 0;
        this.isReady = false;
        this.totalShots = 0;
    }
    
    public Gun() {
        this("Beretta", 8);
    }
    
    public int getAmount() {
        return amount;
    }
    
    public int getCapacity() {
        return capacity;
    }
    
    public boolean ready() {
        return isReady;
    }
    
    public String getModel() {
        return model;
    }
    
    public int getTotalShots() {
        return totalShots;
    }
    
    public void prepare() {
        isReady = !isReady;
    }
    
    public void reload() {
        amount = capacity;
    }
    
    public void shoot() throws NotReadyException, OutOfRoundsException {
        if ( !isReady ) {
            throw new NotReadyException();
        }
        if ( amount == 0 ) {
            throw new OutOfRoundsException();
        }
        
        amount -= 1;
        totalShots += 1;
        isReady = !isReady;
    }
    
    public String toString() {
        return model + " has capacity of " + capacity + " , amount of rounds - " + amount + " and total shoots - " + totalShots;
    }
    
    public Gun clone() throws CloneNotSupportedException {
        return (Gun) super.clone();
    }
}

File: BC-W3/TaskCpp/bcw4/Gun/Gun.cpp
#include "Gun.h"


Gun::Gun(const std::string& model, int capacity) {
    this->model = model;
    this->capacity = capacity;
    this->amount = 0;
    this->totalShots = 0;
    this->isReady = false;
}
Gun::~Gun() {}

int Gun::getAmount() const {
    return this->amount;
}
int Gun::getCapacity() const {
    return this->capacity;
}
bool Gun::ready() const {
    return this->isReady;
}
const std::string& Gun::getModel() const {
    return this->model;
}
int Gun::getTotalShots() const {
    return this->totalShots;
}

void Gun::prepare() {
    this->isReady = !this->isReady;
}
void Gun::reload() {
    this->amount = this->capacity;
}
void Gun::shoot() {
    if ( !this->isReady ) {
        throw NotReady();
    }
    if ( this->amount == 0 ) {
        throw OutOfRounds();
    }
    this->amount -= 1;
    this->totalShots += 1;
    this->isReady = !this->isReady;
}

std::ostream& operator<<(std::ostream& out, const Gun& gun) {
    out << gun.getModel() << " has capacity of " << gun.getCapacity() << " , amount of rounds - " << gun.getAmount() << " and total shoots - " << gun.getTotalShots();
    
    return out;
}